pytorch_lightning/accelerators/gpu_backend.py

```diff
@@ -55,57 +55,27 @@ def train(self):
         return results
 
     def training_step(self, args):
-        if self.trainer.amp_backend == AMPType.NATIVE:
-            with torch.cuda.amp.autocast():
-                output = self.__training_step(args)
-        else:
-            output = self.__training_step(args)
-
-        return output
-
-    def __training_step(self, args):
-        batch = args[0]
-        batch = self.to_device(batch)
-        args[0] = batch
-        output = self.trainer.model.training_step(*args)
-        return output
+        return self.__step(args, 'training_step')
 
     def validation_step(self, args):
-        if self.trainer.amp_backend == AMPType.NATIVE:
-            with torch.cuda.amp.autocast():
-                output = self.__validation_step(args)
-        else:
-            output = self.__validation_step(args)
+        return self.__step(args, 'validation_step')
 
-        return output
+    def test_step(self, args):
+        return self.__step(args, 'test_step')
 
-    def __validation_step(self, args):
-        batch = args[0]
-        batch = self.to_device(batch)
-        args[0] = batch
-        output = self.trainer.model.validation_step(*args)
-        return output
+    def __step(self, args, step_func):
+        args[0] = self.to_device(args[0])
 
-    def test_step(self, args):
         if self.trainer.amp_backend == AMPType.NATIVE:
             with torch.cuda.amp.autocast():
-                output = self.__test_step(args)
+                output = getattr(self.trainer.model, step_func)(*args)
         else:
-            output = self.__test_step(args)
-
-        return output
+            output = getattr(self.trainer.model, step_func)(*args)
 
-    def __test_step(self, args):
-        batch = args[0]
-        batch = self.to_device(batch)
-        args[0] = batch
-        output = self.trainer.model.test_step(*args)
         return output
 
     def to_device(self, batch):
-        gpu_id = 0
-        if isinstance(self.trainer.data_parallel_device_ids, list):
-            gpu_id = self.trainer.data_parallel_device_ids[0]
+        gpu_id = self.trainer.data_parallel_device_ids[0]
```

Comment:
how can gpu_id which is an int be considered as a device here? same thing with horovod backend.

Reply:
Also is this refactor good? Should I do this for all other backends?

Reply:
@awaelchli